### ISSUE TITLE
[93] Grounded Language-Image Pre-training (GLIP)

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#123]
[paper](https://arxiv.org/pdf/2112.03857.pdf)
[code](https://github.com/microsoft/GLIP)

object detection 과 phrase grounding 을 합친 pretraining 을 제안.
코드베이스는 애초에 mask-rcnn 을 이용한 형태.

![image](https://user-images.githubusercontent.com/16400591/175812784-c18c382e-990f-49dc-af85-b7bcac0b7a0f.png)
모델의 zero-shot 결과.

# 선행지식
## Dynamic Head
당시 detection sota 모델 DyHead ([paper](https://arxiv.org/pdf/2106.08322.pdf), [code](https://github.com/microsoft/DynamicHead))
![image](https://user-images.githubusercontent.com/16400591/175813170-230687f2-2933-4246-9281-d9940a9b0d66.png)
![image](https://user-images.githubusercontent.com/16400591/175813204-939fb092-0b68-4037-8858-be7a0d0aa1ae.png)

3가지 attention 으로 이루어진 head 제안.
3가지 attention 을 수식으로 나타내면 다음과 같음
F 는 (L, S, C) 형태이고, L 은 feature pyramid 의 level 개수, S 는 (HxW).
![image](https://user-images.githubusercontent.com/16400591/175813226-55cc8a5f-6467-439c-a6b2-7143f6266f99.png)
### scale-aware attention
어떤 level 의 feature 가 중요한 지 보는 형태.
![image](https://user-images.githubusercontent.com/16400591/175813266-6ea196f6-f52d-4082-a6da-092b303cb0e6.png)

### Spatial-aware Attention
deformable conv 활용, S (HxW) 에 대한 attention 수행.
![image](https://user-images.githubusercontent.com/16400591/175813243-7f693d5a-7188-400d-b004-578cda53028b.png)

### Task-aware Attention
channel-wise attention 수행. (채널을 켜거나 끄는 형태)
![image](https://user-images.githubusercontent.com/16400591/175813252-9bd85a93-9987-4a85-bf27-a4c23a52b30c.png)
![image](https://user-images.githubusercontent.com/16400591/175814035-a2daa006-48e2-4a3e-a05b-86e8bf94aa5a.png)
θ(·) 는 hyperfunction 으로 activation threshold 조정. dynamic relu 와 비슷한 형태로 구현됨. 

# Model
text-encoder 는 bert 를 활용, image-encoder 로는 swin-T, swin-L 활용.
어떠한 detection 모델도 들고 올 수 있는 형태.
caption 이 있는 데이터를 이미지에서 예시로 들고 있음.
명사 부분만 뽑아내서 `". ".join()` 형태로 input 을 만들어 준다.

각 encoder 를 통과시켰다면, fusion layer 를 매번 거쳐서 최대한 feature 간 교환을 시켜주는 형태.
![image](https://user-images.githubusercontent.com/16400591/175812827-abba4107-ff1e-47af-8256-d6e9db6fc474.png)

다음 layer 에는 fusion feature 가 더해진다.
![image](https://user-images.githubusercontent.com/16400591/175814092-8b9bb787-5151-4ceb-b12b-0ef27d7f9d84.png)

fusion 방식은 단순히, dot-product attention 활용하는 형태.
![image](https://user-images.githubusercontent.com/16400591/175814097-3901881a-4fce-43c9-b6c3-b0f97d25d31b.png)
fusion 하면 2배정도 느려짐.
![image](https://user-images.githubusercontent.com/16400591/175815114-5b812e0b-7e0c-494d-968d-6e6b527825dd.png)

최종 feature 들을 갖고 dot-product 를 수행하면, matrix 가 나오게 될 텐데
해당 matrix 에 many-to-one 형태의 cross entropy 가 적용됨. (2-stage 는 CE, 1-stage 는 보통 focal loss)
text sequence 는 명사의 나열이기에 순서에 따른 불이익이 없도록,  bipartite Hungarian match loss 가 사용됨.

localization loss 의 경우 2-stage detector 에서는 rpn loss 등이 될 것이고, 1-stage detector 에서는 centerness loss 등이 될 수 있음.

# Data
꽤나 다양한 데이터가 사용되었음.
### Object Detection Data
- COCO
  - 80 category
  - 이미지 당 5 caption
  - 0.33M
- Objects 365
  - 365 category
  - 2M
- LVIS
  - 1200+ categgory
  - 0.16M
  - 거의 eval 용도.
  - long tail 문제가 있어서 class 별로 균등하게 뽑은 5000장 규모의 minival set 함께 사용

### Grounding Data
 - GoldG (From MDETR)
   - Flickr 30K, VG Caption 등 합친 데이터
   - 0.8M
   - GoldG+ 는 validation 데이터 까지 포함시킨 형태.
 - flickr 30k entites
   - 0.03M

### Caption data
 - Cap (4M)
   - crawling 으로 수집한 데이터
   - pseudo labelling 수행
 - CC (3M)
 - SBU (1M)

# Result
FourODs == Objects365, OpenImages, Visual Genome (excluding COCO images), ImageNetBoxes

![image](https://user-images.githubusercontent.com/16400591/175814663-8c33bdb5-78ff-412e-a997-438c11ed08e2.png)

## COCO zero-shot (detection)
![image](https://user-images.githubusercontent.com/16400591/175814672-7e7165a0-f8be-41d3-b720-249f2d216a9c.png)

## LVIS zero-shot (detection, 1200+ obj)
![image](https://user-images.githubusercontent.com/16400591/175814941-9d4bc0fa-0716-4ca2-a805-462a91fffd75.png)

## Flickr 30K entities (grounding)
![image](https://user-images.githubusercontent.com/16400591/175814914-75d822fb-a9f0-4998-9f2d-dbcb08c200e9.png)

## 다양한 데이터에 따른 결과
![image](https://user-images.githubusercontent.com/16400591/175814960-8c11de16-a087-4e78-8d16-b696f74f9e21.png)

## data efficiency
![image](https://user-images.githubusercontent.com/16400591/175814971-b4a9687d-f89d-4410-a24a-4421bfb4025b.png)

## 다른 dataset 들에 대한 zero-shot 결과들
![image](https://user-images.githubusercontent.com/16400591/175815043-e705a723-2d0c-4dcf-b9ae-4d852e115c9c.png)

## prompt tuning
흥미롭게도, prompt 에 관련 설명을 더 넣어주면 성능이 오름
![image](https://user-images.githubusercontent.com/16400591/175815049-051ec77c-8903-409d-9ea5-4ed63ff43fac.png)
`flat and round` 와 같은 정보를 더 넣어준 것임.

물론 finetuning 하는 게 성능은 더 좋았지만, prompt tuning 도 꽤 괜찮은 성능 향상을 야기했음.
![image](https://user-images.githubusercontent.com/16400591/175815078-a63eca8f-286d-4dae-b535-4d7d71405d8b.png)
